File: cpp/sampleONNXResNet/resnet.cpp
/*=======================================================================
 * Copyright 2020-2023 Enflame. All Rights Reserved.
 *
 *Licensed under the Apache License, Version 2.0 (the "License");
 *you may not use this file except in compliance with the License.
 *You may obtain a copy of the License at
 *
 *http://www.apache.org/licenses/LICENSE-2.0
 *
 *Unless required by applicable law or agreed to in writing, software
 *distributed under the License is distributed on an "AS IS" BASIS,
 *WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *See the License for the specific language governing permissions and
 *limitations under the License.
 *=======================================================================
 */

#include <unistd.h>

#include <algorithm>
#include <cassert>
#include <chrono>
#include <cmath>
#include <cstddef>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <iostream>
#include <vector>

#include "../utils/tops_utils.h"
#include "TopsInference/TopsInferRuntime.h"

char *getCmdOption(char **begin, char **end, const std::string &option) {
  char **itr = std::find(begin, end, option);
  if (itr != end && ++itr != end) {
    return *itr;
  }
  return 0;
}

bool cmdOptionExists(char **begin, char **end, const std::string &option) {
  return std::find(begin, end, option) != end;
}

template <typename T, typename A> int arg_max(std::vector<T, A> const &vec) {
  return static_cast<int>(
      std::distance(vec.begin(), std::max_element(vec.begin(), vec.end())));
}

int main(int argc, char **argv) {
  if (cmdOptionExists(argv, argv + argc, "-h")) {
    std::cout << "usage: resnet [-h]" << std::endl;
    std::cout << "               [-o ONNX MODEL FILE PATH]" << std::endl;
    std::cout << "               [-i PROCESSED IMAGE(224*224*3) FILE PATH]"
              << std::endl;
    std::cout << "               [-n NAME OF MODEL INPUT, DEFAULT IS: input]"
              << std::endl;
    std::cout << "               [-s INPUT SHAPE, DEFAULT IS: (1,3,224,224)]"
              << std::endl;
    return 0;
  }

  const char *onnx_path;
  if (cmdOptionExists(argv, argv + argc, "-o")) {
    onnx_path = getCmdOption(argv, argv + argc, "-o");
  } else {
    std::cout << "[ERROR] Must specify ONNX model path" << std::endl;
    exit(-1);
  }

  const char *img_path;
  if (cmdOptionExists(argv, argv + argc, "-i")) {
    img_path = getCmdOption(argv, argv + argc, "-i");
  } else {
    img_path = "../../data/resnet/ILSVRC2012_val_00006740.JPEG.dat";
  }

  const char *input_names;
  if (cmdOptionExists(argv, argv + argc, "-n")) {
    input_names = getCmdOption(argv, argv + argc, "-n");
  } else {
    input_names = "input";
  }

  const char *input_shapes;
  if (cmdOptionExists(argv, argv + argc, "-s")) {
    input_shapes = getCmdOption(argv, argv + argc, "-s");
  } else {
    input_shapes = "1,3,224,224";
  }

  int precision_type = 2; // 0: default, 1: fp16, 2: mix_fp16
  std::string exec_path =
      engine_name_construct(onnx_path, "../../engines", atoi(input_shapes),
                            get_precision_str(precision_type));

  int card_id = 0;
  // uint32_t cluster_ids[] = {0, 1, 2, 3, 4, 5};  // use all pg to compute
  // uint32_t cluster_num = 6;                     // count of cluster_ids
  uint32_t cluster_ids[] = {0}; // use one pg to compute
  uint32_t cluster_num = 1;     // count of cluster_ids

  int batch_size = 1;

  // 1. init
  TopsInference::topsInference_init();
  void *tops_handler_ =
      TopsInference::set_device(card_id, cluster_ids, cluster_num);

  // 2. load engine file or build it from onnx file
  TopsInference::IEngine *engine = loadOrCreateEngine(
      exec_path.c_str(), onnx_path,
      TopsInference::BuildFlag::TIF_KTYPE_MIX_FP16, input_names, input_shapes);

  // 3. get input&output info from engine
  std::vector<ShapeInfo> inputs_shape_info = get_inputs_shape(engine);
  std::vector<ShapeInfo> outputs_shape_info = get_outputs_shape(engine);

  //  Debug only
  for (auto &s : inputs_shape_info) {
    s.dump();
  }

  for (auto &s : outputs_shape_info) {
    s.dump();
  }

  // 4. alloc host memory for outputs
  std::vector<void *> inputs = alloc_host_memory(inputs_shape_info, batch_size);
  std::vector<void *> outputs =
      alloc_host_memory(outputs_shape_info, batch_size);

  // 5. prepare data
  std::ifstream in_data_stream;
  int data_size = sizeof(float) * 224 * 224 * 3;
  in_data_stream.open(img_path);
  in_data_stream.read(reinterpret_cast<char *>(inputs[0]), data_size);
  in_data_stream.close();
  // int ind = 0;
  // std::ifstream data_list_file("../../data/resnet/test_file.txt");
  // std::string fname, label;
  // while (data_list_file >> fname >> label) {
  //   in_data_stream.open(data_dir + fname);
  //   in_data_stream.read(reinterpret_cast<char *>(inputs[0] + data_size *
  //   ind),
  //                       data_size);
  //   in_data_stream.close();
  //   ind++;
  //   if (ind >= batch_size) break;
  // }
  // data_list_file.close();

  // warmup
  for (int i = 0; i < 3; ++i) {
    engine->run_with_batch(
        batch_size, inputs.data(), outputs.data(),
        TopsInference::BufferType::TIF_ENGINE_RSC_IN_HOST_OUT_HOST);
  }

  // 6. run
  std::chrono::steady_clock::time_point t1 = std::chrono::steady_clock::now();
  auto ret = engine->run_with_batch(
      batch_size, inputs.data(), outputs.data(),
      TopsInference::BufferType::TIF_ENGINE_RSC_IN_HOST_OUT_HOST);
  if (!ret) {
    std::cout << "engine run_with_batch failed." << std::endl;
    exit(-1);
  }
  std::chrono::steady_clock::time_point t2 = std::chrono::steady_clock::now();
  auto time_diff =
      std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count();
  std::cout << "-------------------- " << std::endl;
  std::cout << "gcu inference cost time:" << (int)time_diff << " ms"
            << std::endl;
  std::cout << "-------------------- " << std::endl;

  std::cout << "result:" << std::endl;
  for (int i = 0; i < batch_size; i++) {
    float *start = (float *)outputs[0] + 1000 * i;
    auto v = std::vector<float>(start, start + 1000);
    int max_idx = arg_max(v);
    std::cout << "classification idx=" << max_idx << std::endl;
  }
  std::cout << "-------------------- " << std::endl;

  // 7. free host outputs memory
  free_host_memory(inputs);
  free_host_memory(outputs);

  // 8. release
  TopsInference::release_engine(engine);
  TopsInference::release_device(tops_handler_);
  TopsInference::topsInference_finish();

  return 0;
}

File: python/sampleONNXResNet/resnet.py
#!/usr/bin/env python3
# =======================================================================
# Copyright 2020-2023 Enflame. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# =======================================================================
#

import os
import time
import glob
import argparse
import numbers
from PIL import Image
import numpy as np
import json
import TopsInference


def img_crop(img, top, left, height, width):
    return img.crop((left, top, left + width, top + height))


def img_center_crop(img, output_size):
    if isinstance(output_size, numbers.Number):
        output_size = (int(output_size), int(output_size))
    image_width, image_height = img.size
    crop_height, crop_width = output_size
    crop_top = int(round((image_height - crop_height) / 2.0))
    crop_left = int(round((image_width - crop_width) / 2.0))
    return img_crop(img, crop_top, crop_left, crop_height, crop_width)


def img_resize(img, size, interpolation=Image.BILINEAR):
    if isinstance(size, int):
        w, h = img.size
        if (w <= h and w == size) or (h <= w and h == size):
            return img
        if w < h:
            ow = size
            oh = int(size * h / w)
            return img.resize((ow, oh), interpolation)
        else:
            oh = size
            ow = int(size * w / h)
            return img.resize((ow, oh), interpolation)
    else:
        return img.resize(size, interpolation)


def model_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--device", type=str, default="dtu", help="Which device to run inference"
    )
    parser.add_argument(
        "--card_id",
        type=int,
        default=0,
        help="On which card to run inference processing",
    )
    parser.add_argument(
        "--cluster_ids",
        nargs="+",
        type=int,
        default=0,
        help="On which clusters to run inference. Generally now, \
              cluster_ids is a list with max length 6, it can be \
              [0], [0, 1], [0, 1, 2, 3, 4, 5] or any other range from 0 to 4. \
              cluster_ids can also be conveniently set to -1 to delegate \
              [0, 1, 2, 3, 4, 5]",
    )
    parser.add_argument(
        "--input_names",
        default="input",
        help="Input tensor names, which must be consistent with the content \
        of model file When there are multi input, names are seperated by \
        a comma.",
    )
    parser.add_argument(
        "--input_shape",
        type=str,
        default="1, 3, 224, 224",
        help="The standrd input shape of onnx model",
    )
    parser.add_argument("--data_format", type=str, default="NCHW")
    parser.add_argument(
        "--model_path",
        type=str,
        default="yolov5s.onnx",
        help="Which model to run inference,\
                        should be one of yolov5s, yolov5m, yolov5n, yolov5l, \
                        yolov5x.",
    )
    parser.add_argument("--engine", type=str, help="engine file to be reused.")
    parser.add_argument(
        "--data_path", type=str, help="path to image file or directory(only jpeg now)"
    )
    parser.add_argument(
        "--save_processed_img",
        type=bool,
        default=False,
        help="save the processed numpy image data to the original image dirs",
    )
    parser.add_argument(
        "--output_names",
        default="output",
        help="Output tensor names, which must be consistent with the content \
        of model file When there are multi input, names are seperated by \
        a comma.",
    )
    arg_pars = parser.parse_args()
    return arg_pars


class Resnet:
    time_preprocess = None
    time_infer = None
    time_postprocess = None
    time_total = None

    def __init__(self, arguments, onnx_file, device="dtu"):
        self.device = device
        # dtu device
        self.handler = TopsInference.set_device(
            arguments.card_id, arguments.cluster_ids
        )
        # parse onnx
        parser = TopsInference.create_parser(TopsInference.ONNX_MODEL)
        parser.set_input_names(arguments.input_names)
        parser.set_output_names(arguments.output_names)
        parser.set_input_shapes(arguments.input_shape)
        parser.set_input_dtypes("DT_FLOAT32")

        if arguments.engine is None:
            network = parser.read(onnx_file)
            optimizer = TopsInference.create_optimizer()
            # optimizer.set_build_flag(TopsInference.KFP16)
            # optimizer.set_build_flag(TopsInference.KDEFAULT)
            optimizer.set_build_flag(TopsInference.KFP16_MIX)
            self.engine = optimizer.build(network)
            engine_path = os.path.join(".", os.path.basename(onnx_file) + ".exec")
            self.engine.save_executable(engine_path)
        else:
            self.engine = TopsInference.PyEngine()
            self.engine.load_tops_executable(arguments.engine)

        assert self.engine is not None

        self.input_width = int(arguments.input_shape.split(",")[-1])
        self.input_height = int(arguments.input_shape.split(",")[2])

        self.save_processed_img = arguments.save_processed_img

    def __call__(self, images, images_name):
        inputs = []
        outputs = []
        time_start = time.time()

        for (image, name) in zip(images, images_name):
            processed_img = self.preprocess(image, self.input_width, self.input_height)
            inputs.append(processed_img)
            if self.save_processed_img:
                processed_img.tofile(name + ".dat")

        time_after_preprocess = time.time()
        # Use sync mode, py_stream=None by default
        py_future = self.engine.run_with_batch(
            sample_nums=len(images),
            input_list=np.array([inputs]),
            buffer_type=TopsInference.TIF_ENGINE_RSC_IN_HOST_OUT_HOST,
        )
        pred = py_future.get()
        pred = np.asarray(pred)
        outputs.append(pred)
        time_after_infer = time.time()
        pred = np.argmax(outputs[0], axis=-1)

        time_finish = time.time()
        self.time_preprocess = (time_after_preprocess - time_start) * 1000
        self.time_infer = (time_after_infer - time_after_preprocess) * 1000
        self.time_postprocess = (time_finish - time_after_infer) * 1000
        self.time_total = (time_finish - time_start) * 1000
        return pred

    def preprocess(self, image, input_width, input_height):
        input_size = (input_width, input_height)
        max_size = max(input_width, input_height)

        image = img_resize(image, 256 if max_size <= 256 else 342)
        image = img_center_crop(image, input_size)

        image_data = np.array(image, dtype="float32").transpose(2, 0, 1)
        mean_vec = np.array([0.485, 0.456, 0.406])
        stddev_vec = np.array([0.229, 0.224, 0.225])
        norm_image_data = np.zeros(image_data.shape).astype("float32")
        for i in range(image_data.shape[0]):
            norm_image_data[i, :, :] = (
                image_data[i, :, :] / 255 - mean_vec[i]
            ) / stddev_vec[i]
        norm_image_data = norm_image_data.reshape(3, input_height, input_width).astype(
            "float32"
        )

        return norm_image_data

    def destroy(self):
        TopsInference.release_device(self.handler)


def main(arguments):
    images = []
    images_name = []

    if os.path.isfile(arguments.data_path):
        image_data = Image.open(os.path.join(arguments.data_path)).convert("RGB")
        assert image_data.size[0] > 0, "no image read"
        images.append(image_data)
        images_name.append(os.path.join(arguments.data_path))

    # only jpg
    if os.path.isdir(arguments.data_path):
        jpg_list = glob.glob(arguments.data_path + "/*.JPEG")

        for image in jpg_list:
            image_data = Image.open(os.path.join(arguments.data_path, image)).convert(
                "RGB"
            )
            assert image_data.size[0] > 0, "no image read"
            images.append(image_data)
            images_name.append(os.path.join(arguments.data_path, image))

    with open("imagenet_labels.json", "r") as f:
        labels = json.load(f)

    model_file = arguments.model_path
    classification = Resnet(arguments, onnx_file=model_file, device=arguments.device)

    # warmup
    for i in range(3):
        results = classification(images, images_name)

    results = classification(images, images_name)
    print("{}{}results{}".format("\n" * 2, "*" * 40, "*" * 40))
    for ind, name in enumerate(images_name):
        print(
            "image name: {}, classification: {}".format(name, labels[results[0][ind]])
        )

    time_list = []
    time_list.append(
        [
            classification.time_preprocess,
            classification.time_infer,
            classification.time_postprocess,
            classification.time_total,
        ]
    )

    print("{}{}time cost{}".format("\n" * 2, "*" * 40, "*" * 40))
    print(
        "time_preprocess,{} time_infer,{} time_postprocess,{} time_total".format(
            " " * 3, " " * 10, " " * 5
        )
    )
    for t in time_list:
        print("{}".format(t))


if __name__ == "__main__":
    arguments = model_arguments()
    main(arguments)
